utils/zerovec/src/varzerovec/mod.rs

```diff
@@ -0,0 +1,417 @@
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::str::Utf8Error;
+    /// # use zerovec::VarZeroVecError;
+    /// # use zerovec::VarZeroVec;
+    ///
+    /// let strings = vec!["foo".to_owned(), "bar".to_owned(),
+    ///                    "baz".to_owned(), "quux".to_owned()];
+    /// let bytes = VarZeroVec::get_serializable_bytes(&strings);
+    ///
+    /// let mut vec: VarZeroVec<String> = VarZeroVec::try_from_bytes(&bytes)?;
+    /// assert_eq!(vec.len(), 4);
+    /// # Ok::<(), VarZeroVecError<Utf8Error>>(())
+    /// ```
+    pub fn len(&self) -> usize {
+        match self.0 {
+            VarZeroVecInner::Owned(ref vec) => vec.len(),
+            VarZeroVecInner::Borrowed(components) => components.len(),
+        }
+    }
+    /// Parse a VarZeroVec from a slice of the appropriate format
+    ///
+    /// Slices of the right format can be obtained via VarZeroVec::get_serializable_bytes()
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::str::Utf8Error;
+    /// # use zerovec::VarZeroVecError;
+    /// # use zerovec::VarZeroVec;
+    ///
+    /// let strings = vec!["foo".to_owned(), "bar".to_owned(),
+    ///                    "baz".to_owned(), "quux".to_owned()];
+    /// let bytes = VarZeroVec::get_serializable_bytes(&strings);
+    ///
+    /// let mut vec: VarZeroVec<String> = VarZeroVec::try_from_bytes(&bytes)?;
+    /// assert_eq!(&vec[0], "foo");
+    /// assert_eq!(&vec[1], "bar");
+    /// assert_eq!(&vec[2], "baz");
+    /// assert_eq!(&vec[3], "quux");
+    /// # Ok::<(), VarZeroVecError<Utf8Error>>(())
+    /// ```
+    pub fn try_from_bytes(slice: &'a [u8]) -> Result<Self, ParseErrorFor<T>> {
+        if slice.len() == 0 {
+            // does not allocate
+            return Ok(VarZeroVecInner::Owned(Vec::new()).into());
+        }
+
+        let components = SliceComponents::<T>::new(slice)?;
+
+        Ok(VarZeroVecInner::Borrowed(components).into())
+    }
+
+    /// Obtain an iterator over VarZeroVec's elements
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::str::Utf8Error;
+    /// # use zerovec::VarZeroVecError;
+    /// # use zerovec::VarZeroVec;
+    ///
+    /// let strings = vec!["foo".to_owned(), "bar".to_owned(),
+    ///                    "baz".to_owned(), "quux".to_owned()];
+    /// let bytes = VarZeroVec::get_serializable_bytes(&strings);
+    /// let mut vec: VarZeroVec<String> = VarZeroVec::try_from_bytes(&bytes)?;
+    ///
+    /// let mut iter_results: Vec<&str> = vec.iter().collect();
+    /// assert_eq!(iter_results[0], "foo");
+    /// assert_eq!(iter_results[1], "bar");
+    /// assert_eq!(iter_results[2], "baz");
+    /// assert_eq!(iter_results[3], "quux");
+    /// # Ok::<(), VarZeroVecError<Utf8Error>>(())
+    /// ```
+    pub fn iter<'b: 'a>(&'b self) -> impl Iterator<Item = &'b T::VarULE> {
+        // We use Either here so that we can use `impl Trait` with heterogeneous types
```

Comment:
Praise: Didn't know you could do this.  I might try this trick in various places in regular ZeroVec.

Reply:
hehe :smile: 

it's an annoying limitation that `impl Trait` cannot automagically do this under the hood, but `Either` is decent for this